# Introduction

**Sparrowhawk** is a web platform for doing basic bioinformatic analyses on bacterial genomes locally. It offers genome assembly, sequence mapping and alignment, and taxonomic identification. 

You can see the different sections of this guide in the left, where each of the tabs/methodologies usage is detailed.


## Why use Sparrowhawk?
All analyses are run in your own system, with nothing being uploaded anywhere, nor a job running in a random foreign computing cluster: this makes Sparrowhawk ideal for resource-constrained setups (e.g. bad Internet connection), or for sensitive data studies. The methods are provided as web pages, being thus intuitive to be used, as well as available when you are using a tablet (or even a smartphone!).


<!-- ## Citation -->


## Code repositories
All code used is open source and most of it has been developed inside the [bacterial population genomics group (bacpop)](https://www.bacpop.org) at [EMBL's European Bioinformatics Institute (EMBL-EBI)](https://www.ebi.ac.uk). The main code repositories are the following:

- **Assembly**: [github.com/bacpop/sparrowhawk](https://github.com/bacpop/sparrowhawk).
- **Mapping and alignment**: the software used is ska.rust, whose main repository is [github.com/bacpop/ska.rust](https://github.com/bacpop/ska.rust).
- **Taxonomic identification**: the software used is sketchlib.rust, whose main repository is [github.com/bacpop/sketchlib.rust](https://github.com/bacpop/sketchlib.rust).
- **Website**: [github.com/bacpop/sparrowhawk-web](https://github.com/bacpop/sparrowhawk-web).
